src/libraries/Common/tests/System/Net/Http/HttpAgnosticLoopbackServer.cs

```diff
@@ -0,0 +1,246 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Authentication;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Xunit;
+
+namespace System.Net.Test.Common
+{
+    public class HttpAgnosticLoopbackServer : GenericLoopbackServer, IDisposable
+    {
+        private Socket _listenSocket;
+        private GenericLoopbackOptions _options;
+        private Uri _uri;
+        private List<GenericLoopbackConnection> _connections = new List<GenericLoopbackConnection>();
+
+        public static readonly TimeSpan Timeout = TimeSpan.FromSeconds(30);
+
+        public override Uri Address => _uri;
+
+        public static HttpAgnosticLoopbackServer CreateServer()
+        {
+            return new HttpAgnosticLoopbackServer(new GenericLoopbackOptions());
+        }
+
+        public static HttpAgnosticLoopbackServer CreateServer(GenericLoopbackOptions options)
+        {
+            return new HttpAgnosticLoopbackServer(options);
+        }
+
+        private HttpAgnosticLoopbackServer(GenericLoopbackOptions options)
+        {
+            _options = options;
+            _listenSocket = new Socket(_options.Address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+            _listenSocket.Bind(new IPEndPoint(_options.Address, 0));
+            _listenSocket.Listen(_options.ListenBacklog);
+
+            var localEndPoint = (IPEndPoint)_listenSocket.LocalEndPoint;
+            var host = _options.Address.AddressFamily == AddressFamily.InterNetworkV6 ? $"[{localEndPoint.Address}]" : localEndPoint.Address.ToString();
+            var scheme = _options.UseSsl ? "https" : "http";
+            _uri = new Uri($"{scheme}://{host}:{localEndPoint.Port}/");
+        }
+
+        public override void Dispose()
+        {
+            if (_listenSocket != null)
+            {
+                _listenSocket.Dispose();
+                _listenSocket = null;
+            }
+        }
+        public override async Task<GenericLoopbackConnection> EstablishGenericConnectionAsync()
+        {
+            Socket socket = await _listenSocket.AcceptAsync().ConfigureAwait(false);
+            Stream stream = new NetworkStream(socket, ownsSocket: true);
+
+            if (_options.UseSsl)
```

Comment:
It is weird that this supports non-TLS connections. Servers in the wild do not support this -- when would this be used in our tests?

Reply:
HTTP/1 is still quite common IMHO. And that is what current loopback have as well. As development switch's this would allow to run it without SSL and debug as well is it makes it possible to run on macOS where server ALPN does not work.

Reply:
Slightly changed that so Cory's argument is addressed but the way is not closed completely.

By default for clear text, HTTP/1.1 is expected. This can be overridden via `HttpAgnosticOptions.ClearTextVersion`. You can request specific version (H2) and get an error if an H2 prefix is not received. Or explicitly set `HttpVersion.Unknown` which will do the H2 prefix based negotiation dance I had here originally.